atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/DefaultKeyValueSnapshotReader.java

```diff
@@ -0,0 +1,365 @@
+        return AtlasFutures.getUnchecked(getInternal(
+                (tableRef, cellsByTimestamp) ->
+                        Futures.immediateFuture(transactionKeyValueService.get(tableRef, cellsByTimestamp)),
+                tableReference,
+                cells));
+    }
+
+    @Override
+    public ListenableFuture<Map<Cell, byte[]>> getAsync(TableReference tableReference, Set<Cell> cells) {
+        return getInternal(transactionKeyValueService::getAsync, tableReference, cells);
+    }
+
+    @NotNull
+    private ListenableFuture<Map<Cell, byte[]>> getInternal(
+            CellGetter cellGetter, TableReference tableReference, Set<Cell> cells) {
+        Map<Cell, Long> timestampsByCell = Cells.constantValueMap(cells, startTimestampSupplier.getAsLong());
+        ListenableFuture<Collection<Map.Entry<Cell, byte[]>>> postFilteredResults = Futures.transformAsync(
+                cellGetter.get(tableReference, timestampsByCell),
+                rawResults -> getWithPostFilteringAsync(cellGetter, tableReference, rawResults, Value.GET_VALUE),
+                MoreExecutors.directExecutor());
+
+        return Futures.transform(postFilteredResults, ImmutableMap::copyOf, MoreExecutors.directExecutor());
+    }
+
+    private <T> ListenableFuture<Collection<Map.Entry<Cell, T>>> getWithPostFilteringAsync(
+            CellGetter cellGetter,
+            TableReference tableRef,
+            Map<Cell, Value> rawResults,
+            Function<Value, T> transformer) {
+        long bytes = 0;
+        for (Map.Entry<Cell, Value> entry : rawResults.entrySet()) {
+            bytes += entry.getValue().getContents().length + Cells.getApproxSizeOfCell(entry.getKey());
+        }
+        if (bytes > TransactionConstants.WARN_LEVEL_FOR_QUEUED_BYTES && log.isWarnEnabled()) {
+            log.warn(
+                    "A single get had quite a few bytes: {} for table {}. The number of results was {}. "
+                            + "Enable debug logging for more information.",
+                    SafeArg.of("numBytes", bytes),
+                    LoggingArgs.tableRef(tableRef),
+                    SafeArg.of("numResults", rawResults.size()));
+            if (log.isDebugEnabled()) {
+                log.debug(
+                        "The first 10 results of your request were {}.",
+                        UnsafeArg.of("results", Iterables.limit(rawResults.entrySet(), 10)),
+                        new SafeRuntimeException("This exception and stack trace are provided for debugging purposes"));
+            }
+            //            getHistogram(AtlasDbMetricNames.SNAPSHOT_TRANSACTION_TOO_MANY_BYTES_READ, tableRef)
+            //                    .update(bytes);
+        }
+
+        //        getCounter(AtlasDbMetricNames.SNAPSHOT_TRANSACTION_CELLS_READ, tableRef).inc(rawResults.size());
+
+        Collection<Map.Entry<Cell, T>> resultsAccumulator = new ArrayList<>();
+
+        if (AtlasDbConstants.HIDDEN_TABLES.contains(tableRef)) {
```

Comment:
This is such a weird and underhanded feature, can we remove it?

Reply:
and I think in the world of TransactionKeyValueService actual implementation, I would probably throw

Reply:
We can't remove it just yet - see comment on L153, is used by specific internal workflows. These workflows at least initially won't intersect with our migration work though, and I'm not against having some decorator or similar injection point that bans this case